app/server/lib/HealthChecker.ts

```diff
@@ -0,0 +1,121 @@
+interface PendingServerHealthCheck {
+  expectedCount: number;
+  responses: Record<string, boolean>;
+  resolve: (res: boolean) => void;
+  reject: (err: Error) => void;
+  timeout: NodeJS.Timeout;
+}
+
+export class HealthChecker {
+  private _pendingServerHealthChecks: Map<string, PendingServerHealthCheck>;
+  private _serverInstanceID: string;
+  private _pubSubManager: IPubSubManager;
+
+  constructor(
+    private _server: GristServer
+  ) {
+    this._pubSubManager = createPubSubManager(process.env.REDIS_URL);
+    this._pendingServerHealthChecks = new Map<string, PendingServerHealthCheck>();
+    this._serverInstanceID = process.env.GRIST_INSTANCE_ID || `testInsanceId_${this._server.getHost()}`;
+    this._pubSubManager.getClient()?.sadd('grist-instances', this._serverInstanceID).catch((err) => {
+      log.error('Failed to contact redis', err);
+    });
+    this._subscribeToChannels();
+
+    // Make sure we clean up our Redis mess, if any, even if we exit
+    // by signal.
+    shutdown.addCleanupHandler(null, () => this.close());
+  }
+
+  public async allServersOkay(timeout: number) {
+    const requestId = uuidv4();
+    const client = this._pubSubManager.getClient();
+
+    // If there is no Redis, then our current instance is the only instance
+    const allInstances = await client?.smembers('grist-instances') || [this._serverInstanceID];
+
+    const allInstancesPromise: Promise<boolean> = new Promise((resolve, reject) => {
+      const allInstancesTimeout = setTimeout(() => {
+        log.warn('allServersOkay: timeout waiting for responses');
+        reject(new Error('Timeout waiting for health responses'));
+        this._pendingServerHealthChecks.delete(requestId);
+      }, timeout);
+
+      this._pendingServerHealthChecks.set(requestId, {
+        responses: {},
+        expectedCount: allInstances.length,
+        resolve,
+        reject,
+        timeout: allInstancesTimeout,
+      });
+    });
+    const request: ServerHealthcheckRequest = {
+      id: requestId,
+      checkReady: true
+    };
+    await this._pubSubManager.publish('healthcheck:requests', JSON.stringify(request));
+    return allInstancesPromise;
+  }
+
+  public async close() {
+    await this._pubSubManager.getClient()?.srem('grist-instances', [this._serverInstanceID]);
+    await this._pubSubManager.close();
+  }
+
+  private _subscribeToChannels() {
+    this._pubSubManager.subscribe('healthcheck:requests', async (message) => {
+      const request: ServerHealthcheckRequest = JSON.parse(message);
+      const response: ServerHealthcheckResponse = {
+        instanceId: this._serverInstanceID|| '',
+        requestId: request.id,
+        healthy: !request.checkReady || this._server.ready,
+      };
+      log.debug('allServersOkay request', response);
+      await this._pubSubManager.publish('healthcheck:responses', JSON.stringify(response));
+    });
+
+    this._pubSubManager.subscribe('healthcheck:responses', (message) => {
```

Comment:
If every server subscribes to this, then N servers responding to a request would produce N^2 delivered messages (one for each of the N servers subscribing), I think. Is that anything to worry about?

Reply:
Okay, I've gone back to the dedicated channel idea. That removes the N^2 problem.